experimental/incremental/task.go

```diff
@@ -199,120 +209,106 @@ func (t *Task) aborted() error {
-			for dep := range dep.deps {
-				caller.task.deps[dep] = struct{}{}
-			}
-			if caller.task != nil {
-				dep.downstream.Store(caller.task, struct{}{})
-			}
-		}
-	}()
-
-	// Run the zeroth query synchronously, inheriting this task's semaphore
-	// hold.
-	schedule(0, true)
-
 	if needWait {
 		// Release our current hold on the global semaphore, since we're about to
 		// go to sleep. This avoids potential resource starvation for deeply-nested
 		// queries on low parallelism settings.
 		caller.release()
 		if join.Acquire(caller.ctx, int64(len(queries))) != nil {
-			return nil, context.Cause(caller.ctx)
+			return nil, nil, context.Cause(caller.ctx)
 		}
 
 		// Reacquire from the global semaphore before returning, so
 		// execution of the calling task may resume.
 		if !caller.acquire() {
-			return nil, context.Cause(caller.ctx)
+			return nil, nil, context.Cause(caller.ctx)
 		}
 	}
-
-	return results, context.Cause(caller.ctx)
+	return results, tasks, context.Cause(caller.ctx)
 }
 
 // checkDone returns an error if this task is completed. This is to avoid shenanigans with
 // tasks that escape their scope.
 func (t *Task) checkDone() {
-	if closed(t.result.done) {
+	if t.task.done.Load() {
 		panic("protocompile/incremental: use of Task after the associated Query.Execute call returned")
 	}
 }
 
 // task is book-keeping information for a memoized Task in an Executor.
 type task struct {
-	deps map[*task]struct{} // Transitive.
-
-	// TODO: See the comment on Executor.tasks.
-	downstream sync.Map // [*task, struct{}]
+	// The query that executed this task.
+	query *AnyQuery
 
-	// If this task has not been started yet, this is nil.
-	// Otherwise, if it is complete, result.done will be closed.
+	// This is the sequence ID of the Run call that caused this result to be
+	// computed. If it is equal to the ID of the current Run, it was computed
+	// during the current call. Otherwise, it is cached from a previous Run.
 	//
-	// In other words, if result is non-nil and result.done is not closed, this
-	// task is pending.
-	result atomic.Pointer[result]
+	// Proof of correctness. As long as any Runs are ongoing, it is not possible
+	// for queries to be evicted, so once a query is calculated, its runID is
+	// fixed. Suppose two Runs race the same query. One of them will win as the
+	// leader, and the other will wait until it's done. The leader will mark it
+	// with its run ID, so the leader sees Changed and the loser sees !Changed.
+	// Any other queries from the same or other Runs racing this query will see
+	// the same result.
+	//
+	// Note that runID itself does not require synchronization, because loads of
+	// runID are synchronized-after the done channel being closed.
+	runID uint64
+
+	// The wait group protects the results. All clients must wait but they
+	// may optionally check done to avoid blocking.
+	wg     sync.WaitGroup
+	value  any
+	fatal  error
 	report report.Report
+	done   atomic.Bool
```

Comment:
Is there a posibility of wg and done having false sharing? Best comment on what this atomic is used for.